### PR TITLE
Chat completion notebook with ultrachat_200k dataset

### PR DESCRIPTION
**Pull request by Anubha98**
# Description
Chat completion notebook with ultrachat_200k dataset

# Checklist
Run: https://ml.azure.com/experiments/id/c703ccc6-2b7b-42ac-9371-c25536e688bf/runs/dreamy_nest_d67lhy4440?wsid=/subscriptions/ed2cab61-14cc-4fb3-ac23-d72609214cfd/resourceGroups/training_rg/providers/Microsoft.MachineLearningServices/workspaces/train-finetune-dev-workspace&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

Registered model: https://ml.azure.com/registries/foundation-models-runners-reg/models/Phi-3-mini-4k-instruct-ultrachat-200k/version/1?tid=72f988bf-86f1-41af-91ab-2d7cd011db47

Endpoint Deployment: https://ml.azure.com/endpoints/realtime/aml-shared-model-evaluati-jjdop/test?wsid=/subscriptions/72c03bf3-4e69-41af-9532-dfcdc3eefef4/resourceGroups/shared-model-evaluation-rg/providers/Microsoft.MachineLearningServices/workspaces/aml-shared-model-evaluation-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47

Inference Results (Standard_E8s_v3):
image

- [x] I have read the contribution guidelines.
  - [General](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
  - [SDK](https://github.com/Azure/azureml-examples/blob/main/sdk/CONTRIBUTING.md)
  - [CLI](https://github.com/Azure/azureml-examples/blob/main/cli/CONTRIBUTING.md)
- [x] I have coordinated with the docs team (mldocs@microsoft.com) if this PR deletes files or changes any file names or file extensions.
- [x] Pull request includes test coverage for the included changes.
- [x] This notebook or file is added to the [CODEOWNERS](https://github.com/Azure/azureml-examples/blob/main/.github/CODEOWNERS) file, pointing to the author or the author's team.

### CODE REVIEW

**Review by amangupta26 — DISMISSED** [review]

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+    "except:\n",
+    "    workspace_ml_client = MLClient(\n",
+    "        credential,\n",
+    "        subscription_id=\"<SUBSCRIPTION_ID>\",\n",
+    "        resource_group_name=\"<RESOURCE_GROUP>\",\n",
+    "        workspace_name=\"<WORKSPACE_NAME>\",\n",
+    "    )\n",
+    "\n",
+    "# the models, fine tuning pipelines and environments are available in the AzureML registry, \"models-staging\"\n",
+    "registry_ml_client_staging = MLClient(credential, registry_name=\"models-staging\")\n",
```
Comment (skanakamedal) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:95]: Please create two different mlclients, 1 for fetching the model (from models-staging registry) and other for components (from azureml registry).

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 2]
```diff
@@ -0,0 +1,685 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Chat Completion - Ultrachat-200k\n",
+    "\n",
+    "This sample shows how use `chat-completion` components from the `azureml` system registry to fine tune a model to complete a conversation between 2 people using ultrachat dataset. We then deploy the fine tuned model to an online endpoint for real time inference.\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:10]: Update the description of the dataset.

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+    "\n",
+    "### Training data\n",
+    "We will use the [ultrachat_200k](https://huggingface.co/datasets/HuggingFaceH4/ultrachat_200k) dataset. This dataset is intended to complete conversation between 2 people. with this notebook we will summarize the dialogues and calculate bleu and rouge scores for the summarized text vs provided ground_truth summaries\n",
+    "\n",
+    "### Model\n",
+    "We will use the `Phi-3-mini-4k-instruct` model to show how user can finetune a model for chat-completion task. If you opened this notebook from a specific model card, remember to replace the specific model name. Optionally, if you need to fine tune a model that is available on HuggingFace, but not available in `azureml` system registry, you can either [import](https://github.com/Azure/azureml-examples/blob/main/sdk/python/foundation-models/system/import/import_model_into_registry.ipynb) the model or use the `huggingface_id` parameter instruct the components to pull the model directly from HuggingFace. \n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:16]: Please consider removing the "optional section to remove the support for any custom HF ID"

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2. Pick a foundation model to fine tune\n",
+    "\n",
+    "Phi-3 is a Transformer with 2.7 billion parameters. It was trained using the same data sources as Phi-1.5, augmented with a new data source that consists of various NLP synthetic texts and filtered websites , we need to finetune the model for our specific purpose in order to use it. You can browse these models in the Model Catalog in the AzureML Studio, filtering by the `chat-completion` task. In this example, we use the `Phi-3-mini-4k-instruct` model. If you have opened this notebook for a different model, replace the model name and version accordingly. \n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:109]: Phi-3 model description needs to be updated. It is a 3.8 billion model and not 2.7 billion. The model description can be copied from the model card. Refer to the model card - https://ml.azure.com/registries/models-staging/models/Phi-3-mini-128k-instruct/version/15?flight=itpmerge&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#overview

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+    "    workspace_ml_client = MLClient.from_config(credential=credential)\n",
+    "except:\n",
+    "    workspace_ml_client = MLClient(\n",
+    "        credential,\n",
+    "        subscription_id=\"<SUBSCRIPTION_ID>\",\n",
+    "        resource_group_name=\"<RESOURCE_GROUP>\",\n",
+    "        workspace_name=\"<WORKSPACE_NAME>\",\n",
+    "    )\n",
+    "\n",
+    "# the models, fine tuning pipelines and environments are available in the AzureML registry, \"models-staging\"\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:94]: Can we use _azureml_ registry? instead of _models-staging_? That's where the final model will live afaik.

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 6]
```diff
@@ -0,0 +1,685 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# download the dataset using the helper script. This needs datasets library: https://pypi.org/project/datasets/\n",
+    "import os\n",
+    "\n",
+    "exit_status = os.system(\n",
+    "    \"python ./download-dataset.py --dataset HuggingFaceH4/ultrachat_200k --download_dir ultrachat_200k_dataset --dataset_split_pc 5\"\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:279]: Maybe we can add a description saying that "Running the below script only downloads 5% of the data. you can increase the `dataset_split_pc` to change this parameter"

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+    "            \"content\": \"Certainly! ....\",\n",
+    "            \"role\": \"assistant\"\n",
+    "        }\n",
+    "    ],\n",
+    "    \"prompt_id\": \"d938b65dfe31f05f80eb8572964c6673eddbd68eff3db6bd234d7f1e3b86c2af\"\n",
+    "}\n",
+    "```\n",
+    "\n",
+    "### Model\n",
+    "We will use the [Phi-3-mini-4k-instruct](https://ml.azure.com/registries/models-staging/models/Phi-3-mini-128k-instruct/version/15?flight=itpmerge&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#overview) model to show how user can finetune a model for chat-completion task. If you opened this notebook from a specific model card, remember to replace the specific model name.\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:60]: Is the flight needed? _itpmerge_?

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "[Phi-3-mini-4k-instruct](https://ml.azure.com/registries/models-staging/models/Phi-3-mini-128k-instruct/version/15?flight=itpmerge&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#overview) is a 3.8B parameters, lightweight, state-of-the-art open model built upon datasets used for Phi-2 - synthetic data and filtered websites - with a focus on very high-quality, reasoning dense data. The model belongs to the Phi-3 model family, and the Mini version comes in two variants 4K and 128K which is the context length (in tokens) it can support.\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:62]: repeat: Is the flight needed?

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb (hunk 5 above)
Comment (jpmann) [reply sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:140]: Let's check with Aman on this. We will be publishing this notebook to master only after Phi3 release.

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Chat Completion - Ultrachat-200k\n",
+    "\n",
+    "This sample shows how use `chat-completion` components from the `azureml` system registry to fine tune a model to complete a conversation between 2 people using ultrachat_200k dataset. We then deploy the fine tuned model to an online endpoint for real time inference.\n",
+    "\n",
+    "### Training data\n",
+    "We will use the [ultrachat_200k](https://huggingface.co/datasets/HuggingFaceH4/ultrachat_200k) dataset. This is a heavily filtered version of the UltraChat dataset and was used to train Zephyr-7B-β, a state of the art 7b chat model.\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:13]: We can talk about the training data during the download and visualization section. Just keep the summary here and move the details to the above-mentioned section.

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 10]
```diff
@@ -0,0 +1,731 @@
...
+    "* Removal of dialogues where the assistant replies with phrases like \"I do not have emotions\" or \"I don't have opinions\", even for fact-based prompts that don't involve either.\n",
+    "\n",
+    "**Dataset Structure**\n",
+    "\n",
+    "The dataset has four splits, suitable for:\n",
+    "* Supervised fine-tuning (sft).\n",
+    "* Generation ranking (gen) via techniques like rejection sampling or PPO.\n",
+    "The number of examples per split is shown as follows:\n",
+    "\n",
+    "| train_sft | test_sft | train_gen | test_gen |\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:28]: No need to mention the PPO dataset.

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+    "| :- | :- | :- | :- |\n",
+    "| 207865 | 23110 | 256032 | 28304 |\n",
+    "\n",
+    "\n",
+    "The dataset is stored in parquet format with each entry using the following schema:\n",
```
Comment (jpmann) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:33]: Refer to other notebooks on whether we are adding an example in the markdown section (or) not.

Comment (Anubha98) [reply sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:33]: https://github.com/Azure/azureml-examples/blob/main/sdk/python/foundation-models/system/finetune/translation/translation.ipynb

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb (hunk 7 above)
Comment (Anubha98) [reply sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb]: discussed with @amangupta26 we will be merging the notebook now itself, we can get rid of model card hyperlink for now and have it later once model get published if needed

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb (hunk 8 above)
Comment (Anubha98) [reply sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb]: same as above

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb (hunk 5 above)
Comment (Anubha98) [reply sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:94]: same as above

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+    "except:\n",
+    "    workspace_ml_client = MLClient(\n",
+    "        credential,\n",
+    "        subscription_id=\"<SUBSCRIPTION_ID>\",\n",
+    "        resource_group_name=\"<RESOURCE_GROUP>\",\n",
+    "        workspace_name=\"<WORKSPACE_NAME>\",\n",
+    "    )\n",
+    "\n",
+    "# the models, fine tuning pipelines and environments are available in the AzureML registry, \"azureml\"\n",
+    "registry_ml_client_staging = MLClient(credential, registry_name=\"azureml\")\n",
```
Comment (amangupta26) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:95]: make it `registry_ml_client`

File: sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_name = \"Phi-3-mini-4k-instruct\"\n",
+    "foundation_model = registry_ml_client_staging.models.get(model_name, label=\"latest\")\n",
```
Comment (amangupta26) [inline sdk/python/foundation-models/system/finetune/chat-completion/chat-completion.ipynb:121]: make it `registry_ml_client`

**Review by amangupta26 — APPROVED** [review]

**Review by jpmann — APPROVED** [review]
